e2e/main_test.go

```diff
@@ -0,0 +1,134 @@
+package e2e
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// global client for use by all tests
+var client kubernetes.Interface
+
+// non-configurable for now
+const namespace = "bootkubetesting"
+
+// TestMain handles setup before all tests
+func TestMain(m *testing.M) {
+	var (
+		kubeconfig    = flag.String("kubeconfig", "../hack/quickstart/cluster/auth/kubeconfig", "absolute path to the kubeconfig file")
+		expectedNodes = flag.Int("nodes", 0, "the number of nodes to expect")
+	)
+	flag.Parse()
+
+	if *expectedNodes == 0 {
+		fmt.Println("Please set --nodes flag with the number of nodes in the running cluster")
+		os.Exit(1)
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	// create the clientset
+	client, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	if err := ready(client, *expectedNodes); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	// createNamespace
+	if _, err := createNamespace(client, namespace); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	// run tests
+	exitCode := m.Run()
+
+	if err := deleteNamespace(client, namespace); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	os.Exit(exitCode)
+}
+
+func createNamespace(c kubernetes.Interface, name string) (*v1.Namespace, error) {
+	namespace, err := c.CoreV1().Namespaces().Create(&v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create namespace with name %v %v", name, namespace)
+	}
+	return namespace, nil
+}
+
+func deleteNamespace(c kubernetes.Interface, name string) error {
+	return c.CoreV1().Namespaces().Delete(name, nil)
+}
+
+// Ready blocks until a the cluster is considered available. The current
+// implementation checks that the expected number of nodes are registered.
+func ready(c kubernetes.Interface, expectedNodes int) error {
+	f := func() error {
+		list, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+
+		if len(list.Items) != expectedNodes {
+			return fmt.Errorf("cluster is not ready, expected %v nodes got %v", expectedNodes, len(list.Items))
+		}
+
+		for _, node := range list.Items {
+			for _, condition := range node.Status.Conditions {
+				if condition.Type == v1.NodeReady {
+					if condition.Status != v1.ConditionTrue {
+						return fmt.Errorf("One or more nodes not in the ready state: %v", node.Status.Phase)
+					} else {
+						break
+					}
+				}
+			}
+		}
+
+		return nil
+	}
+
+	if err := retry(12, 10*time.Second, f); err != nil {
+		return err
+	}
+	return nil
+}
+
+func retry(attempts int, delay time.Duration, f func() error) error {
```

Comment:
The `retry` function here will just keep retrying the function until it returns nil. If it runs out of retries it will return the last error returned from the function. So if you were waiting on multiple conditions you'll know which one was the blocking one by writing different errors within the function that is retrying.